Add Ultimate String Guide (#7)
hasufell/hasufell.github.io: posts/2024-05-07-ultimate-string-guide.md

> * The low ten bits (also in the range 0x000–0x3FF) are added to 0xDC00 to give the second 16-bit code unit or low surrogate (W2), which will be in the range 0xDC00–0xDFFF

UTF-16 is not ASCII compatible either. It is more space efficient than UTF-32 though. For some languages, it can
even be more space efficient than UTF-8.

### Unicode Scalar Values

It is important to understand that the Haskell `Char` type (which is essentially a Code Point) can represent surrogates
that are used in UTF-16.

The unicode standard also defines the concept of [Unicode Scalar Values](http://www.unicode.org/versions/Unicode5.2.0/ch03.pdf#page=35):

> Any Unicode code point except high-surrogate and low-surrogate code points. In other words, the ranges of integers 0 to D7FF16 and E00016 to 10FFFF16 inclusive.

So, code point without surrogates. This will become relevant for UTF-8.

### UTF-8

This is similar to UTF-16 a variable-width character encoding. It's often used in web APIs (most notably JSON) and is
often the default on unix systems.

> Here a word is only 8 bit. The number of bytes required depends on the range of the code point and

Reviewer: Perhaps,
```suggestion
Here, a unicode codepoint is represented by a sequence of bytes. The number of bytes required depends on the range of the code point and
```

`String` was already discussed and we don't recommend it, so it's omitted here.

### Text

If you are not sure what you need, you most likely want `Text` from the [text package](https://hackage.haskell.org/package/text),
which is shipped with GHC. This type is meant for human readable Unicode text and has all the primitives you need. The
API is in fact more complete than the one for `String`, containing functions like `stripPrefix` and `toLower`.

Internally, Text uses a UTF-8 encoded byte array since version 2.0 and UTF-16 before version 2.0. So it is always
guaranteed to be valid unicode.

The current definition for strict `Text` is (as of [2.1.1](https://hackage.haskell.org/package/text-2.1.1)):

```hs
-- | A space efficient, packed, unboxed Unicode text type.
data Text = Text
    {-# UNPACK #-} !A.Array -- ^ bytearray encoded as UTF-8
    {-# UNPACK #-} !Int     -- ^ offset in bytes (not in Char!), pointing to a start of UTF-8 sequence
    {-# UNPACK #-} !Int     -- ^ length in bytes (not in Char!), pointing to an end of UTF-8 sequence
```

As we can see here, this type allows efficient slicing to avoid unnecessary `memcpy` for many operations.

> E.g. `init` and `tail` are *O(1)* time and space. `splitAt` is *O(1)*, but *O(n)* time, because UTF-8 complicates
> the offset computation.

Reviewer: ```suggestion
E.g. `init` and `tail` are *O(1)* time and space. `splitAt` is not *O(1)*, but *O(n)* time, because UTF-8 complicates
the offset computation.
```

It is better to use a proper streaming library.

## Streaming

Streaming can not only solve the lazy IO problem, but may also
solve some of the inefficiency of the `[Char]` type, while keeping a similarly
simple API.

There are many popular streaming libraries. A few of them are:

- [conduit](https://hackage.haskell.org/package/conduit)
- [streaming](https://hackage.haskell.org/package/streaming)
- [streamly](https://hackage.haskell.org/package/streamly)
- [pipes](https://hackage.haskell.org/package/pipes)

### Via Streamly

A couple of years ago I wrote the blog post
[From conduit to streamly](https://hasufell.github.io/posts/2021-10-22-conduit-to-streamly.html),
which gives an introduction into both streamly and conduit. The streamly API
has diverged quite a bit since then, with multiple major versions. So I won't
go into much detail about it.

However, streamly is one notable example which provides an alternative
to the `[Char]` type in [Streamly.Unicode.Stream](https://hackage.haskell.org/package/streamly-core-0.2.2/docs/Streamly-Unicode-Stream.html):

```hs
decodeUtf8 :: Monad m => Stream m Word8 -> Stream m Char
encodeUtf8 :: Monad m => Stream m Char -> Stream m Word8
```

> A very simple program to print the last unicode char of a file via streamly is:

Reviewer: last unicode char? The example shows first unicode char.

```hs
import System.Environment (getArgs)
import Streamly.Data.Stream (Stream)
import Streamly.Data.Fold (Fold)
import qualified Streamly.Data.Fold as Fold
import qualified Streamly.FileSystem.File as File
import qualified Streamly.Unicode.Stream as Unicode

main :: IO ()
main = do
  (file:_) <- getArgs
  c <- getFirstCharFromFile file
  print c

getFirstCharFromFile :: FilePath -> IO (Maybe Char)
getFirstCharFromFile file = stream `Fold.drive` fold
 where
  stream :: Stream IO Char
  stream = Unicode.decodeUtf8Chunks $ File.readChunks file

  fold :: Monad m => Fold m a (Maybe a)

>   fold = Fold.one

Reviewer: `Fold.latest`?
`streamly-string` uses the correct implementation although named `getFirstCharFromFile`.

> * The low ten bits (also in the range 0x000–0x3FF) are added to 0xDC00 to give the second 16-bit code unit or low surrogate (W2), which will be in the range 0xDC00–0xDFFF

UTF-16 is not ASCII compatible either. It is more space efficient than UTF-32 though. For some languages, it can
even be more space efficient than UTF-8.

### Unicode Scalar Values

It is important to understand that the Haskell `Char` type (which is essentially a Code Point) can represent surrogates
that are used in UTF-16.

The unicode standard also defines the concept of [Unicode Scalar Values](http://www.unicode.org/versions/Unicode5.2.0/ch03.pdf#page=35):

> Any Unicode code point except high-surrogate and low-surrogate code points. In other words, the ranges of integers 0 to D7FF16 and E00016 to 10FFFF16 inclusive.

So, code point without surrogates. This will become relevant for UTF-8.

### UTF-8

This is similar to UTF-16 a variable-width character encoding. It's often used in web APIs (most notably JSON) and is
often the default on unix systems.

> Here a word is only 8 bit. The number of bytes required depends on the range of the code point and

Reviewer 2: "word"?

Author: It's actually "code unit", but I found it confusing, so was looking for a simpler expression.

- UTF-8, UTF-16 and UTF-32 are text encodings with different trade offs
- surrogates are a special case for UTF-16 (`Unicode Scalar Values = Unicode Code Points - surrotages`)

Going back to the definition of "character", we now see the confusion:

- a surrogate can hardly be a visible character
- the visible character  🇯🇵 needs two code points to be expressed (and there are many others)

This has lead to yet another definition: **"Grapheme Cluster"**. This is specified by the [Unicode Standard Annex #29](https://www.unicode.org/reports/tr29/), which deals with determining boundaries between characters, words and sentences.
It is, again, quite technical, but is much closer to "user visible character".

## Back to Haskell String type

Now that we know what a **Unicode Code Point** is, we also understand that the Haskell String type
has essentially no *text encoding*. It is just a linked list of those code points (a subset of `Int`, in fact).
This can be a nice property, e.g. as an intermediate representation when converting between encodings (say UTF-8 to UTF-16).

However, it is a **questionable default for a String type**, because:

> - it is inefficient for large text (carries the overhead of a linked list with thunks for every `Char`)

Reviewer 3: It might be useful to copy a diagram from page 2 of https://github.com/Bodigrim/my-talks/blob/master/zurihac2022/slides.pdf, demonstrating how wasteful `String` is.

(this is also discussed in [Fixing 'FilePath' in Haskell](https://hasufell.github.io/posts/2022-06-29-fixing-haskell-filepaths.html)).
An alternative type is `ShortByteString`, which will be discussed next.

#### ByteString summary

Invariants:

- pinned memory

Useful for:

- large data
- dealing with raw bytes (e.g. web servers)
- dealing with C FFI
- storing non-unicode encodings e.g. via a newtype wrapper
- fast parsers, see the excellent blog post from Chris Done on [Fast Haskell: Competing with C at parsing XML](https://chrisdone.com/posts/fast-haskell-c-parsing-xml/)

Not so useful for:

- dealing with unicode or human readable text
- dealing with lots of small byte sequences

Lazy variants, again, are useful for streaming and incremental processing, as the strict variant requires the whole content to be in memory.

### ShortByteString

This type is from the bytestring package as well and lives under
[Data.ByteString.Short](https://hackage.haskell.org/package/bytestring-0.12.1.0/docs/Data-ByteString-Short.html).

It has the same API as `ByteString` since [0.11.3.0](https://hackage.haskell.org/package/bytestring-0.11.3.0/changelog),

> so can be used as a drop-in replacement. The main difference is that it is *unpinned memory*, so causes no

Reviewer 3: This is not strictly correct: `ByteArray` could be allocated as pinned. 
```suggestion
so can be used as a drop-in replacement. The main difference is that it could be backed by *unpinned memory*, so causes no
```

Author: However, using the default API to construct ShortByteString's, we always end up with unpinned ones, no?

Reviewer 3: Yes

heap fragmentation. It also has no lazy variant.

The definition as of [0.12.1.0](https://hackage.haskell.org/package/bytestring-0.12.1.0) is:

```hs
newtype ShortByteString =
  ShortByteString
  { unShortByteString :: ByteArray
  }
```

This makes it suitable for things like unix filepaths. But we will explore better filepath types later.

The name is maybe a little bit misleading. It can very well be used for large data as well, if you
don't mind its strictness (the whole content is always in memory). **However, this type does not allow slicing**,
unlike `Text` and `ByteString`, and so a lot of operations cause `memcpy`.

Interfacing with C FFI triggers memory copy as well, because we need pinned memory.

#### ShortByteString summary

Invariants:

- unpinned memory

Useful for:

- lots of small to medium sized byte sequences
- large data, if strictness is desired and efficient slicing not needed
- dealing with C FFI (although it incurs `memcpy`)
- storing non-unicode encodings e.g. via a newtype wrapper

Not so useful for:

- dealing with unicode or human readable text

> - fast parsers, because no lazy variant and no efficient slicing

Reviewer 3: It would be helpful to underline availability of efficient slicing for other types.

Author: Into how much detail should I go there? ByteString slicing is more efficient than Text (e.g. `splitAt`), but overall they are more efficient than Lists.

Author: I would go for **Text**:

> As we can see here, this type allows efficient slicing to avoid unnecessary `memcpy` for many operations.
E.g. `init` and `tail` are *O(1)* time and space. `splitAt` is *O(1)*, but *O(n)* time, because UTF-8 complicates
the offset computation.

And for **ByteString**:

> This allows, similar to Text, slicing without copying memory (through pointer arithmetic and the length field).
Since we're not dealing with unicode, but just `Word8` boundaries, operations like `splitAt` are *O(1)* time and
space.

And then consider Text *quite* efficient and ByteString *very* efficient for slicing.

Reviewer 3: You probably meant "is O(1) __space__, but O(n) time"
